### ISSUE TITLE
Hitting 10K unsummarized ops results in client getting into infinite loop of reconnects

### ISSUE THREAD

**Issue by vladsud** [microsoft/FluidFramework#7137]
From looking at session in TAP40 scalability runs, it's pretty clear that when client gets to 10K unsummarized ops, we get into this look that results in continuous disconnects and reconnects, with exactly same reasons:

fluid:telemetry:Container:ConnectionStateChange_Disconnected | reason = "Nack: Submit a summary before inserting additional operations"
fluid:telemetry:Container:ContainerWarning | error = "Service busy/throttled: Nack: Submit a summary before inserting additional operations"

There has to be some mechanism not to get to such infinite loop. Either by injecting delays, or better - having read-only connection that we do not attempt to convert to "write" connection until problem is fixed. Note that these nacks are happening in "write" connection, and client does not go through "read" connection cycle because we have non-acked op (from prior attempts), so logic in DM always forces us on a path of "write" reconnect. So if we implement some solution that allows us to keep this connection to listen for summary ack, we need to ensure "connected" event is not exposed to runtime that will attempt to send an op again, or there is some buffering logic in DM that does not actually sends this op to socket to avoid getting nack. Or we ensure that nack does not result in disconnect and we keep that connection until we learn about new summary.

**Comment by vladsud** [comment]
@jatgarg , @anthony-murphy - FYI, feel free to share your thoughts / suggestions.

**Comment by heliocliu** [comment]
We observed this for some documents in the wild as well.  They were eventually able to successfully submit a summary (seen in both telemetry and in the op stream), but continued the reconnect loop because the server kept nack'ing with the too many ops error.  End result was that the op stream after that first nack error was entirely join/leave/noClient ops (tens of thousands of them too).  So if this gets addressed with the read-only connection and wait for summary op fix, some sort of delays/throttling for reconnect loops is likely necessary too to guard against weird server behavior like this.

**Comment by vladsud** [comment]
@heliocliu , what you suggest below sounds like a PUSH bug (on top of runtime bugs), If that's the case, can you please work with @GaryWilber to get to the bottom of it? This bug should not be used for that, if we believe there is server bug, separate ticket should be opened against PUSH. If that's some kind of client bug, we should get crisper on what is the bug and also open separate issue for that.

**Comment by anthony-murphy** [comment]
The more control we can pass into the runtime the easier something like this would be to control. right now summaries and nacks are at very different layers which makes it really hard coordinate the two without further increasing coupling, and complexity. The runtime is in a much better position to do something like make itself logically read-only, and wait for a summary ack to reenable writes. this seems like a good case to start moving delta manger logic into the runtime layer.

**Comment by heliocliu** [comment]
The PUSH issue is being tracked in ICM and has begun to be addressed, starting with #7166.  It just seemed weird in this case that the client continued reconnecting immediately even though it was never really able to keep the connection, though it does require bad behavior from the server.  So maybe not in the client's scope of responsibility.

**Comment by vladsud** [comment]
We should consider
- adding UTs to cover this case
- inject a failure into stress tests that stalls summarizer and tests recovery for this path.

**Comment by vladsud** [comment]
We should also consider https://github.com/microsoft/FluidFramework/issues/7589 - a case where 10K unsummarized nack happens when actually there are very few unsummarized ops - can client have recovery mechanism that engages last summary flow in such case even though current logic skips it?

**Comment by vladsud** [comment]
So I was looking at it from angle of - add some safety net at DeltaManager level, as we know there could be many reasons when we get into infinite nack loop.
The interesting tidbits are:
- We can't use total number of nack events in session to close container - there are a lot of sessions that see "Nack: Nonexistent client" at 5 min interval (there is an issue tracking this to not get into that state, but it just shows we can't blindly use count). Similar for "Nack: Readonly client" - there are sessions with a lot these (spaced out, of course). That said, in both cases we do see some other reason (not nack) of disconnect in between such nacks. So we could close container if we do see a lot of nacks with no other disconnects in some reasonable timeframe.
- The max rate I see for nacks is at 10 second frequency. I'm a bit surprised here, as I'd expect whole pipeline to get into much faster cadence here (with existing logic).

So something like that might be feasible as a safety net
If client sees 3+ nacks in 60 second timeframe, and nacks are the only reasons for reconnects, then terminate client.
60 seconds gives is long enough time frame to given client to recover if it's recoverable situation.
This query also captures nacks in this proportion: 99% are 10K unsummarized ops, 1% of readonly and non-existing client.

We can also go with 2+ nacks in 30 second interval - it has somewhat similar properties.

I think it would be much better to see if simple rule like 2 nacks in a row (without other disconnect reasons) works well, but I can't figure out how to construct such qurry :)

**Comment by vladsud** [comment]
https://github.com/microsoft/FluidFramework/pull/7753 is a pre-work for this item. I need it to hit production and observe that indeed we have very few nacks and all of them can be treated as catastrophic events if they repeat couple times. Once this info is confirmed, I can make simple fix here.

So moving to Nov.

**Comment by vladsud** [comment]
I think solution here should be to handle all nacks as critical container errors.
I eliminated most of the workflows that relies on nacks. When you look at telemetry:

> Office_Fluid_FluidRuntime_Performance
| where Event_Time > ago(5d)
| where Data_eventName contains "_Disconnect"
| where Data_loaderVersion contains "0.5"
| where Data_reason contains "Nack"
| summarize count() by Data_reason, Data_loaderVersion

You see this picture:
| Data_reason	| Data_loaderVersion	| count_ |
|-|-|-|
| Nack (BadRequestError): Nonexistent client	| 0.54.2 | 784 |
| Nack (BadRequestError): Nonexistent client	| 0.54.1 | 550 |
| Nack (BadRequestError): Nonexistent client	| 0.53.0 | 93 |
| Nack (ThrottlingError): Submit a summary before inserting additional operations | 0.52.1 | 55 |
| Nack (BadRequestError): Nonexistent client	| 0.52.1 | 2402 |
| Nack (ThrottlingError): Submit a summary before inserting additional operations | 0.50.4	173 |
| Nack (BadRequestError): Nonexistent client | 0.50.4 | 144 |

In other words, there are only two types of nacks:
- summary related that should close container. We should double check that it does not stop last summarizer from continuing with summary
- "Nonexistent client" issues. That's https://github.com/microsoft/FluidFramework/issues/8483 - it should be fixed first (and validated through telemetry that these types of nacks are gone) before we can start working on this item.

**Comment by vladsud** [comment]
Here is what I'd suggest doing:
- I no longer think that getting rid of all nacks is possible, at least due to a race condition that I've described in #8483. I think we should still reduce them to a minimum and understand the system, but we should decouple that issue from this issue.
- For this issue, we should look in debugger for the right way to identify it (ideally not based on text of error messages, ideally based on unique error code that ordering service sends us), and when we receive this nack, close container.

Key pointer - getNackReconnectInfo(), that's where we process nacks and decide if the bug is retryable or not.

How to repro - The easiest way is to do two things (local code changes):
- disable summaries altogether. Search for disableSummaries
- Add 10K ops, either manually (hard), or just right some code that will keep sending ops, let's say 10 ops/s (note - if you send too fast, ODSP can throttle client and disconnect for other reason).

Once file has 10K ops, you should see this nack,

**Comment by vladsud** [comment]
I believe this will be addressed (though not directly) by https://github.com/microsoft/FluidFramework/pull/9243 and we likely can close this item.

**Comment by vladsud** [comment]
Resolved by above mentioned PR
